### PR TITLE
Change CSSLint to Stylelint

### PR DESCRIPTION
**Pull request by rei-nert**
Change CSSLint as CSS lint to use Stylelint.
Solve #55  and #46 
Test passed just like using CSSLint

### CODE REVIEW

File: server/cssfile.js [hunk 1]
```diff
@@ -35,18 +35,21 @@ CssFile.prototype.selectorFromPosition = function(line, column){
...
+			callback(errors);
+			return;
 		}
-	});
 
-	if(errors.length > 0 && callback){
-		callback(errors);
-		return;
-	}
+	})
```
Comment (turbio) [inline server/cssfile.js:51]: I think it'd be a good idea to throw a `.catch(...)` on the end in case something goes wrong, don't wanna be left hanging.

File: server/filemanager.js [hunk 2]
```diff
@@ -30,7 +30,7 @@ FileManager.prototype.newFile = function(id, name, path, type, source){
 			//+ ', path: ' + path + ', type: ' + type);
 
 	if(source == undefined){
-		source = '';
```
Comment (turbio) [inline server/filemanager.js:33]: what's the purpose of this change?

**Review by turbio — CHANGES_REQUESTED** [review]
Awesome! Thanks for contributing. 

PR looks really good to me, I'll get it merged in soon as I get a chance to test it locally.

File: server/filemanager.js (hunk 2 above)
Comment (rei-nert) [reply server/filemanager.js:33]: Well, as I was reviewing it now, I think I made this change thinking that the source variable was the path to the .css file, and I changed it because of the stylelint documentation:

"_If the directory of cacheLocation does not exist, make sure you add a trailing / on *nix systems or \ on Windows. Otherwise, stylelint assumes the path to be a file._"

But I guess I misunderstood the variable, as it's the css code itself.

File: server/cssfile.js (hunk 1 above)
Comment (rei-nert) [reply server/cssfile.js:51]: Yes, but I haven't thought how to do it. 

I thought about using the "errored" property of the stylint

"
_errored

Boolean. If true, at least one rule with an "error"-level severity registered a violation._
"
[Node.js API of Stylelint](https://stylelint.io/user-guide/usage/node-api)

I'm not that much experienced with Stylelint. I just did a little research to implement it and solve the bug I was experiencing.

File: server/filemanager.js (hunk 2 above)
Comment (turbio) [reply server/filemanager.js:33]: gotcha, sounds like it'd be a good idea to revert this change then.

**Comment by rei-nert** [comment]
Changed stylelint promise to have a catch and fix a bug where parsed the CSS with lint errors, by making it a promise and using Promise.all() to have both fulfilled.

**Comment by jdsutherland** [comment]
I just checked out this PR and tried a few different files and they all errored after starting Bracey. I hope this can get worked at as Bracey is often unusable in files containing modern CSS features.

**Comment by rei-nert** [comment]
> I just checked out this PR and tried a few different files and they all errored after starting Bracey. Switching to Stylelint seems like it should resolve the many issues related to Bracey blowing up with files containing modern CSS features, so I hope this PR works out.

Can you post the errors? All the tests passed, and I also checked with a few of my files and everything was working fine.
If you can post the log, I'd gladly try to solve the errors

**Comment by jdsutherland** [comment]
Sorry, @rei-nert. I ran `npm install` forgetting that `run npm install --prefix server` required, so the problem was that `stylelint` was missing. I tried this again and it's working great. This should resolve #36.

File: server/cssfile.js [hunk 3]
```diff
@@ -35,42 +35,64 @@ CssFile.prototype.selectorFromPosition = function(line, column){
...
+	}).catch(err => {
+		return err
+	})
 
-	if(errors.length > 0 && callback){
-		callback(errors);
-		return;
+	const parser =  new Promise((res, rej) => { 
+		try {
+			const parsed = cssparser.parse(source)
```
Comment (turbio) [inline server/cssfile.js:60]: since cssparser is already a simple blocking call I don't think it should be wrapped in a promise.

File: server/cssfile.js [hunk 4]
```diff
@@ -35,42 +35,64 @@ CssFile.prototype.selectorFromPosition = function(line, column){
 };
 
 CssFile.prototype.setContent = function(source, callback){
-	var messages = csslint.verify(source).messages;
-	var errors = [];
-	messages.forEach(function(msg){
-		if(msg.type == 'error'){
-			errors.push(msg);
+	const errors = [];
```
Comment (turbio) [inline server/cssfile.js:38]: Having two concurrent promises is neat but passing one result through the function scoped var `errors` and the other through the promise return needs to be cleaned up a bit. Lets move cssparser out of a promise (as mentioned above there's not point, it's already a simple blocking call) and clean up the stylelint call to not need this errors var.

File: server/cssfile.js [hunk 5]
```diff
@@ -35,42 +35,64 @@ CssFile.prototype.selectorFromPosition = function(line, column){
...
-		return;
+	Promise.all([lint, parser]).then(result => {
+		var changed = (this.source != undefined && this.source != source);
+		if(changed){
+			callback(null);
+		}else{
+			callback(null, null);
+		}
 	}
+	).catch(err => {
```
Comment (turbio) [inline server/cssfile.js:86]: my understanding of js promises is this `.catch` should never be called since the error will already be caught in each of promise `.catch`es above. I'm suspicious this is a fluke of the test suite. What's actually happening here is the callback on line 80/83 is being called incorrectly during the test which throws an error from `expect(err).to.not.be.null` which then leads to this `.catch` and seemingly correct behavior. You can fix the test suite by having it assert the error value and not just the fact that it errored.

File: server/cssfile.js [hunk 6]
```diff
@@ -35,42 +35,64 @@ CssFile.prototype.selectorFromPosition = function(line, column){
...
+			if(msg.errored){
+				msg.warnings.forEach(warning => {
+					errors.push(warning)
+				});
+			}
+		})
+		return results
+	}).then(results => {
+		if(errors.length > 0 && callback){
+			throw new Error("Errors found")
```
Comment (turbio) [inline server/cssfile.js:51]: lets avoid magic error strings in code :sparkles:

**Review by turbio — CHANGES_REQUESTED** [review]
apologies for the long review time, finally got a chance to take a look. I think we should focus on simplifying the code. There are some particularly spooky behavior where tests should be failing but are instead passing due to a quirk of the test suite.

File: server/cssfile.js (hunk 3 above)
Comment (rei-nert) [reply server/cssfile.js:60]: I tried not wrapping it in a promise, but because stylelint uses a promise.

So, if I don't use a promise in the cssparsed, it'll always try to parse the code, even with lint giving an error.

This would raise exceptions, because it would try to parse a wrong css code.

Putting it in an promise delay its execution, and allow stylelint to error, having a callback with the errors, as it was before

Comment (turbio) [reply server/cssfile.js:60]: My bad, I should have been more clear. You can just call `cssparser.parse` inside the `.then` callback, There's no need to create the second promise. This also has the advantage of not running the css parser when you don't have to.

Comment (rei-nert) [reply server/cssfile.js:60]: I tried this, but because of the `this.parsed = cssparser.parse(source)` and the `this.parsed` call in selectorFromPosition function, using `cssparser.parse` in the `.then` would give an error, because `this.parsed` would be `undefined`

Comment (turbio) [reply server/cssfile.js:60]: I'm having a bit of trouble understanding why `this.parsed` being undefined would be an error. It shouldn't matter if `this.parsed` is undefined since it's only being set. It might be that perhaps `this` is undefined if you're not using an arrow function???

Comment (rei-nert) [reply server/cssfile.js:60]: No, the error is in the function `selectorFromPosition`, because it uses `this.parsed` in its for loop, so if it's undefined, it'll raise an exception

Putting the cssparsed in the `.then` would make delay its execution, making the `selectedFromPosition` to try to interact with an undefined variable

But when I put the css parsed in its own promise, it resolves this problem

Comment (turbio) [reply server/cssfile.js:60]: Ah gotcha! It'd be best to match the previous behavior which leaves `this.parsed` undefined when there is a lint or parse error. As long as `setContent` matches the previous behavior we shouldn't have to worry about it. The previous behavior was to leave `this.parsed` undefined so lets do that that! For reference I believe the previous goal of the linter was as a safeguard the parse function, I'm suspicious the linting rules may have accidentally changed when moving to the new linter.

Side note: Sounds like `selectorFromPosition` erring when `this.parsed` is undefined is a separate bug we can address in another PR. As long as we match the previous behavior this shouldn't matter.

File: server/.stylelintrc.json [hunk 7]
```diff
@@ -0,0 +1,3 @@
+{
```
Comment (turbio) [inline server/.stylelintrc.json:1]: unclear exactly how/when this file is read by the library. It'd be better if this could be embedded into the code to avoid spooky action at a distance

Comment (rei-nert) [reply server/.stylelintrc.json:1]: Looking in the Stylelint documentation, there is:

> stylelint uses cosmiconfig to find and load your configuration object. Starting from the current working directory, it looks for the following possible sources:
> -  a stylelint property in package.json
>  -  a .stylelintrc file
>  -  a stylelint.config.js file exporting a JS object
>  - a stylelint.config.cjs file exporting a JS object. When running stylelint in JavaScript packages that specify "type":"module" in their package.json  


> No rules are turned on by default and there are no default values. You must explicitly configure each rule to turn it on.

> You can extend an existing configuration (whether your own or a third-party one).
> 
> Popular configurations include:
> 
>     stylelint-config-recommended - turns on just possible error rules
>     stylelint-config-standard - extends recommended one by turning on 60 stylistic rules

I used the standard configuration because I don't have much time to create a file with all the rules, and it'd be better to use the library standard config file

**Comment by rei-nert** [comment]
Undo stylistic changes and undo promise around CSS parser, returning to previous behavior.
Also, I found that stylelint has a plugin to css parser and I'm trying to make it work

**Review by basicer — APPROVED** [review]

**Comment by noahgitsham** [comment]
Why was this never merged?
Could this project be adopted/forked by someone else if it is abandoned?

I'm considering writing a clone in lua for neovim.

**Comment by turbio** [comment]
heyo @noahgitsham! IIRC this PR never got to state where it addressed all the comments and replicated existing behavior closely enough. Haven't really found the time to give this plugin the love it deserves and fix a lot of the outstanding bugs. Plus the codes a bit kludgey and architectures absolutely bizarre (originally wrote this in highschool :upside_down_face:). I'm super open to supporting any work around this plugin (e.g. add you as a collaborator or direct to a fork/clone) so lemme know (shoot me an email or hmu on discord @ turbio).

**Comment by noahgitsham** [comment]
@turbio Hey, it's lovely to hear that you haven't completely abandoned ship. This plugin is very impressive considering you wrote it in high school!
Admittedly I have written very little vimscript to be of any use to this project. While I like the idea of support for both vim and neovim, I'm not sure about the direction that Bram is taking vim in. I would personally much prefer to write in lua if I were to develop a plugin. I'll have a look through your code, but it's your decision if you wish to restart maintenance on this project.

**Comment by turbio** [comment]
Thanks @noahgitsham. Big agree on a pure lua rewrite, I've flirted with the idea before but haven't found the motivation. The architecture ended up this way partly because of the insanity that is vimscript + vim's language bindings so I'm right there with you. Using python bindings + nodejs was by far the most regrettable decision and has led to endless problems distributing and developing this plugin. Everything inside neovim without all the moving parts sounds awesome.

I've no plans to maintain this repo in its current state beyond reviewing pull requests unless I find the motivation for a rewrite. If you have any technical questions let me know, always down to chat.

**Comment by noahgitsham** [comment]
@turbio cool, in that case I think I'll start making something and see where it goes. I've never worked on something like this, so I'll be learning as I go along. I'm imagining it will turn out as poorly written if not worse than your project 😅.

It's been great talking to you though, I'll almost certainly be in contact with any questions
